api/users.go

```diff
@@ -53,35 +60,383 @@ func (self *ApiServer) SetPassword(
 	if err != nil {
 		return nil, Status(self.verbose, err)
 	}
 
-	permissions := acls.READ_RESULTS
-	perm, err := acls.CheckAccess(org_config_obj, user_record.Name, permissions)
+	perm, err := acls.CheckAccess(org_config_obj, user_record.Name, acls.SERVER_ADMIN)
 	if !perm || err != nil {
 		return nil, status.Error(codes.PermissionDenied,
 			"User is not allowed to enumerate users.")
 	}
 
+	filterOrgs := true
+
+	if isMemberOfRootOrg(user_record.Orgs) {
+		filterOrgs = false
+	} else {
+		ok, err := acls.CheckAccess(org_config_obj, user_record.Name, acls.ORG_ADMIN)
+		if err != nil {
+			return nil, err
+		}
+		if ok {
+			filterOrgs = false
+		}
+	}
+
 	result := &api_proto.Users{}
 
 	users, err := users_manager.ListUsers(ctx)
 	if err != nil {
 		return nil, Status(self.verbose, err)
 	}
 
+	if filterOrgs {
+		newUsers := []*api_proto.VelociraptorUser{}
+		callerOrgs := map[string]bool{}
+		for _, org := range user_record.Orgs {
+			callerOrgs[org.Id] = true
+
+		}
+
+		for _, user := range users {
+			for _, org := range user.Orgs {
+				_, found := callerOrgs[org.Id]
+				if found {
+					newUsers = append(newUsers, user)
+					break
+				}
+			}
+		}
+		users = newUsers
+	}
+
+	if in.WithRoles {
+		for _, user := range users {
+			user.Permissions, err = acls.GetPolicy(org_config_obj, user.Name)
+			if err != nil {
+				return nil, err
+			}
+		}
+	}
+
+	sort.Sort(userByName(users))
 	result.Users = users
 
 	return result, nil
 }
 
+func checkOrgMembership(requestedOrgs []*api_proto.Org, hasOrgs []*api_proto.Org) bool {
+	if isMemberOfRootOrg(hasOrgs) {
+		return true
+	}
+
+	for _, requestedOrg := range requestedOrgs {
+		found := false
+		for _, hasOrg := range hasOrgs {
+			if utils.CompareOrgIds(requestedOrg.Id, hasOrg.Id) {
+				found = true
+				break
+			}
+		}
+
+		if !found {
+			return false
+		}
+	}
+
+	return true
+}
+
+func setPassword(org_config_obj *config_proto.Config,
+		 user_record *api_proto.VelociraptorUser, password string,
+		 createUser bool) error {
+	authenticator, err := authenticators.NewAuthenticator(org_config_obj)
+	if err != nil {
+		return err
+	}
+
+	// Set a random password if none is provided to prevent login should the authentication
+	// method is later changed to one that does require a password
+	if authenticator.IsPasswordLess() && password == "" &&
+	   (createUser || len(user_record.PasswordHash) == 0) {
+		randomPassword := make([]byte, 100)
+		_, err = rand.Read(randomPassword)
+		if err != nil {
+			return err
+		}
+		password = string(password)
+	}
+
+	if createUser && password == "" {
+		// Do not accept an empty password if we are using a password
+		// based authenticator.
+		return fmt.Errorf("Authentication requires a password but one was not provided.")
+	}
+
+	if password != "" {
+		users.SetPassword(user_record, password)
+	}
+
+	return nil
+}
+
+func grantUserRoles(config_obj *config_proto.Config, org_manager *services.OrgManager,
```

Comment:
Should we just call the VQL function https://github.com/Velocidex/velociraptor/blob/master/vql/server/users/grant.go#L23

It makes it much easier to test and we then know the logic is the same as what is tested.